src/contracts/SmartContract.sol

```diff
@@ -140,6 +140,13 @@ contract SmartContract {
     function addSolution(bytes calldata solution, bytes calldata witness) public {
         require(contractStatus == Status.Solving, "Contract is not in Solving status.");
         require(nextUnsolvedPuzzlePart < amountOfPuzzleParts, "All puzzle parts have already been solved.");
+        require(checkSolution(solution, witness, puzzleParts[nextUnsolvedPuzzlePart].commitment), "The solution is not correct.");
+
+        uint256 upperBound = puzzleParts[nextUnsolvedPuzzlePart].upperBound;
+        require(block.timestamp <= initialTimestamp + upperBound, "Too late: the time upper bound has been exceeded.");
+
+        // Once the solution is correct, the solver should be paid
+        pay(nextUnsolvedPuzzlePart);
```

Comment:
this makes sense here, but it deviates from the specification:

>PayDTLP(aj , adrTPH, coins, j) → bj . This phase involves SC. It can be invoked by C or S.
• if aj = 1, then:
a) if bj̸ = 1, for j-th puzzle, send coinsj coins to
T PH, where coins j ∈ coins.
b) set bj = 1. This ensures that one will not be
paid multiple times for the same solution.
• if aj = 0, then send coinsj coins back to S and
set bj = 0.


That's not the end of the world, though. For consistency though, when is payBack called?

Reply:
The payback  is called when the helper wants to return the money back if no one has solved the puzzle in time

Reply:
What would be better to make it consistent?